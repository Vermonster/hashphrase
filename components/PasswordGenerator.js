import React from 'react';
import { Keyboard, Clipboard, View } from 'react-native';
import loplop from 'loplop';
import FormInput from './FormInput';
import BaseButton from './BaseButton';

export default class PasswordGenerator extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      label: '',
      password: '',
    };
  }

    handleChange = name => (e) => {
      const inputValue = e.nativeEvent.text;
      this.setState({ [name]: inputValue });
    };

    addToClipboard = (password) => {
      const { onClipboardSave } = this.props;
      Clipboard.setString(password);
      onClipboardSave();
    };

    generatePassword = (e) => {
      e.preventDefault();
      const { label, password } = this.state;
      this.setState({ generatedPassword: loplop(label, password) }, () => {
        const { generatedPassword } = this.state;
        this.addToClipboard(generatedPassword);
      });
      Keyboard.dismiss();
    };

    render() {
      const { label, password } = this.state;
      return (
        <View>
          <FormInput name="label" value={label} placeholder="Label..." onChanged={this.handleChange('label')} />
          <FormInput name="password" value={password} placeholder="Master Password..." secureTextEntry onChanged={this.handleChange('password')} />
          <BaseButton onPress={this.generatePassword} buttonText="Generate Password" />
        </View>
      );
    }
}
