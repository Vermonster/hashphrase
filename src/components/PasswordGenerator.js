import React from 'react';
import {
  Keyboard, Clipboard, View, StyleSheet, Text,
} from 'react-native';
import { withNamespaces } from 'react-i18next';
import { Button, IconButton } from 'react-native-paper';
import loplop from 'loplop';
import ConfirmationDialog from './ConfirmationDialog';
import FormInput from './FormInput';
import { colors, rowCenter } from '../styles/base';

const styles = StyleSheet.create({
  generatorContainer: {
    justifyContent: 'center',
    alignItems: 'center',
  },
  warning: {
    flexDirection: 'row',
    width: '80%',
    justifyContent: 'space-between',
    marginTop: '-10%',
  },
  icon: {
    marginHorizontal: 0,
  },
});

class PasswordGenerator extends React.Component {
  state = {
    label: '',
    password: '',
    confirmPassword: '',
    generatedPassword: '',
    disabled: true,
    modalVisibility: false,
    inputError: false,
  };

  componentDidUpdate(prevProps) {
    const { isNewPassword } = this.props;
    if (isNewPassword !== prevProps.isNewPassword) {
      this.handleSubmitButtonState();
    }
  }

  showModal = () => this.setState({ modalVisibility: true });

  closeModal = () => this.setState({ modalVisibility: false });

  handleSubmit = (e) => {
    e.preventDefault();
    const { label, password, confirmPassword } = this.state;
    if ((password && confirmPassword) && (password !== confirmPassword)) {
      this.setState({ inputError: true });
    } else {
      this.setState({ inputError: false }, this.generatePassword(label, password));
    }
  };

  addToClipboard = (password) => {
    Clipboard.setString(password);
  };

  handleChange = name => (e) => {
    const inputValue = e.nativeEvent.text;
    this.setState({ [name]: inputValue, inputError: false }, this.handleSubmitButtonState);
  }

  handleSubmitButtonState = () => {
    this.setState({ disabled: true });
    const { label, password, confirmPassword } = this.state;
    const { isNewPassword } = this.props;
    return ((isNewPassword && label && password && confirmPassword)
    || (!isNewPassword && label && password)) ? this.setState({ disabled: false }) : null;
  }

  generatePassword(type, password) {
    this.setState({ generatedPassword: loplop(type, password) }, () => {
      const { generatedPassword } = this.state;
      this.addToClipboard(generatedPassword);
    });
    Keyboard.dismiss();
    this.showModal();
  }

  confirmEntry(inputType) {
    const { label, password, confirmPassword } = this.state;
    let status = null;
    if (inputType === 'label') {
      status = (label.length > 0);
    } else if (inputType === 'password') {
      status = (password.length > 0);
    } else if (inputType === 'confirmPassword') {
      status = (password === confirmPassword);
    }
    this.setState((prevProps) => {
      
    });
    return status;
  }

  render() {
    const {
      generatedPassword, disabled, modalVisibility, inputError, label, password, confirmPassword,
    } = this.state;
    const { isNewPassword, showSnackbar, t } = this.props;

    return (
      <View style={styles.generatorContainer}>
        <FormInput
          value={label}
          componentType="label"
          placeholderText={t('label')}
          prompt={t('passwordLabel')}
          label={t('label')}
          handleChange={this.handleChange}
          onBlur={this.confirmEntry}
        />
        <FormInput
          value={password}
          componentType="password"
          placeholderText={t('password')}
          prompt={t('masterPassword')}
          label={t('password')}
          handleChange={this.handleChange}
        />
        { isNewPassword && (
          <>
            <FormInput
              value={confirmPassword}
              componentType="confirmPassword"
              placeholderText={t('confirmPassword')}
              prompt={t('retypePassword')}
              label={t('confirmPassword')}
              error={inputError}
              handleChange={this.handleChange}
              style={styles.textInput}
            />
            { inputError && (
              <View style={[rowCenter, styles.warning]}>
                <Text style={{ color: colors.warning }}>{t('warning')}</Text>
                <IconButton icon="warning" color={colors.warning} style={styles.icon} />
              </View>
            )
            }
          </>
        )
      }
        <Button
          testID="submit-create-password"
          onPress={this.handleSubmit}
          accessibilityLabel="CREATE ACCOUNT PASSWORD BUTTON"
          disabled={disabled}
          mode="contained"
          dark
        >
          {t('submitMessage')}
        </Button>
        <ConfirmationDialog
          visible={modalVisibility}
          generatedPassword={generatedPassword}
          closeModal={this.closeModal}
          showSnackbar={showSnackbar}
        />
      </View>
    );
  }
}

export default withNamespaces('passwordGenerator')(PasswordGenerator);

export { PasswordGenerator as TestPasswordGenerator };
