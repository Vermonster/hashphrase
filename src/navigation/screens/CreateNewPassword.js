import React, { Component } from 'react';
import { Appbar } from 'react-native-paper';
import { View } from 'react-native';
import PasswordGenerator from '../../../components/PasswordGenerator';
import BaseNotification from '../../../components/BaseNotification';

class CreateNewPassword extends Component {
  constructor(props) {
    super(props);
    this.state = {
      showCopy: false,
    };
  }

  showNotification = () => {
    this.setState({ showCopy: true });
  }

  render() {
    const copy = 'Generated password was copied to the clipboard';
    const { showCopy } = this.state;

    return (
      <View>
        <Appbar.Header>
          <Appbar.Content title="Placeholder file for a future component" />
        </Appbar.Header>
        <PasswordGenerator onClipboardSave={this.showNotification} />
        { showCopy && <BaseNotification message={copy} /> }
      </View>
    );
  }
}

export default CreateNewPassword;
