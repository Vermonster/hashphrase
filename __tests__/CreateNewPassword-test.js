import React from 'react';
import { shallow } from 'enzyme';
import { View } from 'react-native';
import { Appbar } from 'react-native-paper';
import CreateNewPassword from '../src/navigation/screens/CreateNewPassword';
import PasswordGenerator from '../components/PasswordGenerator';

describe('/CreateNewPassword />', () => {
  it('should render the same snapshot', () => {
    const wrapper = shallow(<CreateNewPassword />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should contain all of the required components', () => {
    const wrapper = shallow(<CreateNewPassword />);
    const requiredComponents = [
      Appbar.Header,
      Appbar.Content,
      View,
      PasswordGenerator,
    ];

    requiredComponents.forEach((component) => {
      expect(wrapper.find(component)).toHaveLength(1);
    });
  });

  it('should toggle state when showNotification is called', () => {
    const wrapper = shallow(<CreateNewPassword />);
    const instance = wrapper.instance();
    expect(wrapper.state().showCopy).toEqual(false);
    instance.showNotification();
    expect(wrapper.state().showCopy).toEqual(true);
  });
});
