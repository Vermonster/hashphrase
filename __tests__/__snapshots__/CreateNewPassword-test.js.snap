// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/CreateNewPassword /> should render the same snapshot 1`] = `
<Component>
  <withTheme(Appbar.Header)>
    <withTheme(Appbar.Content)
      title="Placeholder file for a future component"
    />
  </withTheme(Appbar.Header)>
  <PasswordGenerator
    onClipboardSave={[Function]}
  />
</Component>
`;
